pina/solvers/pinns/rbapinn.py

```diff
@@ -0,0 +1,140 @@
+
+    where :math:`r_i` denotes the residual at point :math:`i`, 
+    math:`\gamma` denotes the decay rate, and :math:`\eta` is
+    the learning rate for the weights' update.
+
+    .. seealso::
+        **Original reference**: Sokratis J. Anagnostopoulos, Juan D. Toscano,
+        Nikolaos Stergiopulos, and George E. Karniadakis.
+        "Residual-based attention and connection to information 
+        bottleneck theory in PINNs".
+        Computer Methods in Applied Mechanics and Engineering 
+        421 (2024): 116805.
+        DOI: `10.1016/
+        j.cma.2024.116805 <https://doi.org/10.1016/j.cma.2024.116805>`_.
+    """
+
+    def __init__(
+        self,
+        problem,
+        model,
+        extra_features=None,
+        loss=torch.nn.MSELoss(),
+        optimizer=torch.optim.Adam,
+        optimizer_kwargs={"lr": 0.001},
+        scheduler=ConstantLR,
+        scheduler_kwargs={"factor": 1, "total_iters": 0},
+        eta = 0.001,
+        gamma = 0.999,
+    ):
+        """
+        :param AbstractProblem problem: The formulation of the problem.
+        :param torch.nn.Module model: The neural network model to use.
+        :param torch.nn.Module extra_features: The additional input
+            features to use as augmented input.
+        :param torch.nn.Module loss: The loss function used as minimizer,
+            default :class:`torch.nn.MSELoss`.
+        :param torch.optim.Optimizer optimizer: The neural network optimizer to
+            use; default is :class:`torch.optim.Adam`.
+        :param dict optimizer_kwargs: Optimizer constructor keyword args.
+        :param torch.optim.LRScheduler scheduler: Learning
+            rate scheduler.
+        :param dict scheduler_kwargs: LR scheduler constructor keyword args.
+        :param float eta: The learning rate for the weights of the residual.
+        :param float gamma: The decay parameter in the update of the weights
+            of the residual.
+        """
+        super().__init__(
+            problem=problem,
+            model=model,
+            extra_features=extra_features,
+            loss=loss,
+            optimizer=optimizer,
+            optimizer_kwargs=optimizer_kwargs,
+            scheduler=scheduler,
+            scheduler_kwargs=scheduler_kwargs,
+        )
+
+        self.eta = eta
+        self.gamma = gamma
+        self.weights = {}
+        for condition_name in problem.conditions:
+            self.weights[condition_name] = 0
+
+
+    def loss_phys(self, samples, equation):
+        """
+        Computes the physics loss for the residual-based attention PINN 
+        solver based on given samples and equation.
+
+        :param LabelTensor samples: The samples to evaluate the physics loss.
+        :param EquationInterface equation: The governing equation
+            representing the physics.
+        :return: The physics loss calculated based on given
+            samples and equation.
+        :rtype: LabelTensor
+        """
```

Comment:
Everything is fine! The only thing is that when using `self.store_log` for software consistency you should store the **unweighted** residual.

You can easily accomplish this by defining on the init a `_vectorial_loss` (like [here](https://github.com/mathLab/PINA/blob/7634d22ca7c6b80bd594096a9f131bc19454e7fa/pina/solvers/pinns/sapinn.py#L200)), and then rewrite the loss evaluation as:

```
  residual = self.compute_residual(samples=samples, equation=equation)
  cond = self.current_condition_name

  r_norm = self.eta*torch.abs(residual)/torch.max(torch.abs(residual))
  self.weights[cond] = (self.gamma*self.weights[cond] + r_norm).detach()

  loss_value = self._vectorial_loss(
      torch.zeros_like(residual, requires_grad=True), residual
  )
  self.store_log(loss_value=float(self._vect_to_scalar(loss_value)))
  return self._vect_to_scalar(self.weights[cond]*loss_value)
```
where:
```
def _vect_to_scalar(self, loss_value):
        """
        Elaboration of the pointwise loss through the mask model and the
        self adaptive weights

        :param LabelTensor loss_value: the matrix of pointwise loss

        :return: the scalar loss
        :rtype LabelTensor
        """
        if self.loss.reduction == "mean":
            ret = torch.mean(loss_value)
        elif self.loss.reduction == "sum":
            ret = torch.sum(loss_value)
        else:
            raise RuntimeError(
                f"Invalid reduction, got {self.loss.reduction} "
                "but expected mean or sum."
            )
        return ret
```
I think with this everything should work. What do you think?

Reply:
It sounds good to me. However, I believe that the weights should be squared when returning the value of the weighted loss, as follows:

```python
return self._vect_to_scalar(self.weights[cond]**2*loss_value)
```

Indeed, in the previously submitted version weights were inside the MSE, which involves squaring. 
I totally agree on everything else.

Reply:
Perfecto for the square!